proposals/mainnet_migration.md

```diff
@@ -0,0 +1,528 @@
+
+```python
+addDPoSModuleEntry():
+    DPoSObj = {}
+    DPoSObj.validators = createValidatorsArray()
+    DPoSObj.voters = createVotersArray()
+    DPoSObj.snapshots = []
+    DPoSObj.genesisData = createGenesisDataObj()
+    data = serialization of DPoSObj using genesisDPoSStoreSchema
+    append {"moduleID": MODULE_ID_DPOS, "data": data} to b.assets
+
+createValidatorsArray():
+    validators = []
+    validatorKeys = getValidatorKeys()
+    for every account in accounts:
+        if account.dpos.delegate.username == "":
+            continue
+        validator = {}
+        validator.address = account.address
+        validator.name = account.dpos.delegate.username
+        validator.blsKey = INVALID_BLS_KEY
+        validator.proofOfPossession = DUMMY_PROOF_OF_POSSESSION
+        if validatorKeys[account.address]:
+            validator.generatorKey = validatorKeys[account.address]
+        else:
+            validator.generatorKey = INVALID_ED25519_KEY
+        validator.lastGeneratedHeight = account.dpos.delegate.lastForgedHeight
+        validator.isBanned = account.dpos.delegate.isBanned
+        validator.pomHeights = account.dpos.delegate.pomHeights
+        validator.consecutiveMissedBlocks = account.dpos.delegate.consecutiveMissedBlocks
+        validators.append(validator)
+    return validators
+
+# This functions gets the public keys of the registered validators,
+# i.e., the accounts for which account.dpos.delegate.username is not
+# the empty string, from the history of Lisk Core v3 blocks and
+# transactions.
+getValidatorKeys():
+    let keys be an empty dictionary
+    for every block c with height in [HEIGHT_PREVIOUS_SNAPSHOT_BLOCK + 1,  HEIGHT_SNAPSHOT]:
+        address = address(c.generatorPublicKey)
+        keys[address] = c.generatorPublicKey
+        for trs in c.transactions:
+            address = address(trs.senderPublicKey)
+            if the validator corresponding to address is a registered validator:
+                keys[address] = trs.senderPublicKey
+    return keys
+
+createVotersArray():
+    voters = []
+    for every account in accounts:
+        if account.dpos.sentVotes == [] and account.dpos.unlocking == []:
+            continue
+        voter = {}
+        voter.address = account.address
+        voter.sentVotes = account.dpos.sentVotes
+        voter.pendingUnlocks = account.dpos.unlocking
+        voters.append(voter)
+    return voters
+
+createGenesisDataObj():
+    genesisDataObj = {}
+    genesisDataObj.initRounds = DPOS_INIT_ROUNDS
+    r = round number for the block a
+    Let topDelegates be the set of the top 101 non-banned delegate accounts by delegate weight at the end of round r-2
+    initDelegates = [account.address for account in topDelegates]
+    sort initDelegates in lexicographical order
+genesisDataObj.initDelegates = initDelegates
+return genesisDataObj
```

Comment:
The indentation of the two lines above is incorrect.

Reply:
Fixed!